File: backend/src/integrations/kms/kms_metadata.js
/**
 * KMS & HSM Key Metadata Model & Invariants — Phase 14.3
 *
 * Enforces the 7 required metadata dimensions:
 * 1. key identifier
 * 2. algorithm
 * 3. size/parameters
 * 4. state
 * 5. rotation metadata
 * 6. owner
 * 7. usage metadata
 *
 * CRITICAL SECURITY INVARIANT:
 * "Never extract protected private key material."
 * Enforces strict runtime assertion that private key bytes, exponents, seeds,
 * or raw secret material are NEVER populated or returned.
 */

// Forbidden field names that must NEVER appear in metadata
const FORBIDDEN_PRIVATE_KEY_FIELDS = new Set([
  "privatekey",
  "private_key",
  "d",
  "p",
  "q",
  "dp",
  "dq",
  "qi",
  "rawkey",
  "raw_key",
  "secretbytes",
  "secret_bytes",
  "secretmaterial",
  "secret_material",
  "privatekeybytes",
  "private_key_bytes",
  "seed",
  "masterkey",
  "master_key",
  "keymaterial",
  "key_material",
]);

class ProtectedKeyMaterialError extends Error {
  constructor(message) {
    super(message);
    this.name = "ProtectedKeyMaterialError";
  }
}

class KmsKeyMetadata {
  /**
   * @param {Object} options
   * @param {string} options.keyId - Canonical key identifier (ARN, URI, ID, path)
   * @param {string} options.algorithm - Cryptographic algorithm (AES-256-GCM, RSA-2048, etc.)
   * @param {number|string} options.size - Key size in bits (256, 2048, etc.) or parameter spec
   * @param {string} options.state - Lifecycle state (Enabled, Disabled, Active, etc.)
   * @param {Object} options.rotation - Rotation metadata { enabled, periodDays, lastRotatedAt, nextRotationAt, version }
   * @param {string} options.owner - Account, project, Vault namespace, or Owner tag
   * @param {Object} options.usage - Usage metadata { keyUsage, operations, origin, isExportable }
   * @param {string} [options.provider] - 'aws_kms' | 'gcp_kms' | 'azure_keyvault' | 'hashicorp_vault' | 'pkcs11_hsm'
   * @param {string} [options.description] - Description or label
   * @param {Object} [options.rawMetadata] - Raw vendor metadata with private fields scrubbed
   */
  constructor({
    keyId,
    algorithm,
    size,
    state,
    rotation = {},
    owner,
    usage = {},
    provider = "unknown",
    description = "",
    rawMetadata = {},
  }) {
    if (!keyId) throw new Error("KmsKeyMetadata requires non-empty 'keyId'");
    if (!algorithm) throw new Error("KmsKeyMetadata requires non-empty 'algorithm'");
    if (size === undefined || size === null) throw new Error("KmsKeyMetadata requires 'size' (key length or parameters)");
    if (!state) throw new Error("KmsKeyMetadata requires non-empty 'state'");
    if (!owner) throw new Error("KmsKeyMetadata requires non-empty 'owner'");

    // 1. Enforce strict invariant: Never allow private key material
    KmsKeyMetadata.assertNoPrivateKeyMaterial(rawMetadata);

    this.keyId = String(keyId);
    this.algorithm = String(algorithm).toUpperCase();
    this.size = typeof size === "number" ? size : String(size);
    this.state = String(state);
    this.provider = String(provider).toLowerCase();
    this.description = String(description || "");
    this.owner = String(owner);

    // 2. Rotation metadata
    this.rotation = Object.freeze({
      enabled: Boolean(rotation.enabled),
      periodDays: rotation.periodDays !== undefined ? Number(rotation.periodDays) : null,
      lastRotatedAt: rotation.lastRotatedAt ? String(rotation.lastRotatedAt) : null,
      nextRotationAt: rotation.nextRotationAt ? String(rotation.nextRotationAt) : null,
      version: rotation.version !== undefined ? String(rotation.version) : "1",
    });

    // 3. Usage metadata
    this.usage = Object.freeze({
      keyUsage: String(usage.keyUsage || "ENCRYPT_DECRYPT"),
      operations: Array.isArray(usage.operations) ? [...usage.operations] : ["encrypt", "decrypt"],
      origin: String(usage.origin || "KMS"),
      isExportable: Boolean(usage.isExportable), // Typically false for KMS/HSM
    });

    this.rawMetadata = Object.freeze(KmsKeyMetadata.sanitizeMetadata(rawMetadata));

    Object.freeze(this);
  }

  /**
   * Runtime guard enforcing zero private key material.
   */
  static assertNoPrivateKeyMaterial(obj, path = "") {
    if (!obj || typeof obj !== "object") return;

    for (const key of Object.keys(obj)) {
      const normalized = key.toLowerCase().replace(/[-_]/g, "");
      if (FORBIDDEN_PRIVATE_KEY_FIELDS.has(normalized)) {
        throw new ProtectedKeyMaterialError(
          `CRITICAL SECURITY VIOLATION: Private key material detected in field '${path ? `${path}.${key}` : key}'. ECDAT collects metadata only!`
        );
      }
      if (typeof obj[key] === "object" && obj[key] !== null) {
        this.assertNoPrivateKeyMaterial(obj[key], path ? `${path}.${key}` : key);
      }
    }
  }

  /**
   * Recursively sanitizes raw metadata to guarantee safety.
   */
  static sanitizeMetadata(obj) {
    if (!obj || typeof obj !== "object") return {};
    const clean = {};
    for (const [k, v] of Object.entries(obj)) {
      const normalized = k.toLowerCase().replace(/[-_]/g, "");
      if (FORBIDDEN_PRIVATE_KEY_FIELDS.has(normalized)) {
        clean[k] = "[REDACTED_PROTECTED_MATERIAL]";
      } else if (typeof v === "object" && v !== null && !Array.isArray(v)) {
        clean[k] = this.sanitizeMetadata(v);
      } else {
        clean[k] = v;
      }
    }
    return clean;
  }

  /**
   * Converts KMS key metadata into CycloneDX CBOM component.
   */
  toCbomComponent() {
    return {
      type: "cryptographic-asset",
      "bom-ref": `cbom:${this.provider}:${this.keyId}`,
      name: this.keyId,
      description: this.description || `${this.provider.toUpperCase()} Managed Key (${this.algorithm})`,
      cryptoProperties: {
        assetType: "key",
        algorithm: this.algorithm,
        keyLength: typeof this.size === "number" ? this.size : parseInt(this.size, 10) || 0,
        state: this.state,
        provider: this.provider,
        rotationEnabled: this.rotation.enabled,
        rotationPeriodDays: this.rotation.periodDays,
        owner: this.owner,
        keyUsage: this.usage.keyUsage,
        isExportable: this.usage.isExportable,
      },
      properties: [
        { name: "ecdat:kms:provider", value: this.provider },
        { name: "ecdat:kms:key_id", value: this.keyId },
        { name: "ecdat:kms:owner", value: this.owner },
        { name: "ecdat:kms:rotation_enabled", value: String(this.rotation.enabled) },
        { name: "ecdat:kms:state", value: this.state },
      ],
    };
  }

  toJSON() {
    return {
      keyId: this.keyId,
      algorithm: this.algorithm,
      size: this.size,
      state: this.state,
      rotation: this.rotation,
      owner: this.owner,
      usage: this.usage,
      provider: this.provider,
      description: this.description,
      rawMetadata: this.rawMetadata,
    };
  }
}

module.exports = {
  KmsKeyMetadata,
  ProtectedKeyMaterialError,
  FORBIDDEN_PRIVATE_KEY_FIELDS,
};

File: scanners/integrations/kms/kms_metadata.py
"""
KMS & HSM Key Metadata Model & Invariants — Phase 14.3

Enforces the 7 required metadata dimensions:
1. key identifier
2. algorithm
3. size/parameters
4. state
5. rotation metadata
6. owner
7. usage metadata

CRITICAL SECURITY INVARIANT:
"Never extract protected private key material."
Enforces strict runtime assertions prohibiting private key bytes, exponents, seeds,
or raw secret material.
"""

from typing import Dict, Any, Optional, Union, List
import copy

FORBIDDEN_PRIVATE_KEY_FIELDS = frozenset(
    {
        "privatekey",
        "private_key",
        "d",
        "p",
        "q",
        "dp",
        "dq",
        "qi",
        "rawkey",
        "raw_key",
        "secretbytes",
        "secret_bytes",
        "secretmaterial",
        "secret_material",
        "privatekeybytes",
        "private_key_bytes",
        "seed",
        "masterkey",
        "master_key",
        "keymaterial",
        "key_material",
    }
)


class ProtectedKeyMaterialError(Exception):
    """Raised when any protected private or raw secret key material is detected."""

    pass


def assert_no_private_key_material(obj: Any, path: str = "") -> None:
    """Runtime invariant guard ensuring zero private key material is present."""
    if not isinstance(obj, dict):
        return

    for key, value in obj.items():
        normalized = str(key).lower().replace("-", "").replace("_", "")
        if normalized in FORBIDDEN_PRIVATE_KEY_FIELDS:
            full_path = f"{path}.{key}" if path else str(key)
            raise ProtectedKeyMaterialError(
                f"CRITICAL SECURITY VIOLATION: Private key material detected in field '{full_path}'. "
                "ECDAT collects metadata only!"
            )
        if isinstance(value, dict):
            assert_no_private_key_material(value, f"{path}.{key}" if path else str(key))
        elif isinstance(value, (list, tuple)):
            for idx, item in enumerate(value):
                if isinstance(item, dict):
                    assert_no_private_key_material(item, f"{path}.{key}[{idx}]")


def sanitize_metadata(obj: Any) -> Any:
    """Recursively redacts forbidden private key fields."""
    if not isinstance(obj, dict):
        return obj

    cleaned = {}
    for k, v in obj.items():
        normalized = str(k).lower().replace("-", "").replace("_", "")
        if normalized in FORBIDDEN_PRIVATE_KEY_FIELDS:
            cleaned[k] = "[REDACTED_PROTECTED_MATERIAL]"
        elif isinstance(v, dict):
            cleaned[k] = sanitize_metadata(v)
        elif isinstance(v, list):
            cleaned[k] = [sanitize_metadata(item) if isinstance(item, dict) else item for item in v]
        else:
            cleaned[k] = v
    return cleaned


class KmsKeyMetadata:
    """Standardized metadata representation for KMS and HSM managed keys."""

    def __init__(
        self,
        key_id: str,
        algorithm: str,
        size: Union[int, str],
        state: str,
        owner: str,
        rotation: Optional[Dict[str, Any]] = None,
        usage: Optional[Dict[str, Any]] = None,
        provider: str = "unknown",
        description: str = "",
        raw_metadata: Optional[Dict[str, Any]] = None,
    ):
        if not key_id:
            raise ValueError("KmsKeyMetadata requires non-empty 'key_id'")
        if not algorithm:
            raise ValueError("KmsKeyMetadata requires non-empty 'algorithm'")
        if size is None:
            raise ValueError("KmsKeyMetadata requires 'size' (key length or curve parameters)")
        if not state:
            raise ValueError("KmsKeyMetadata requires non-empty 'state'")
        if not owner:
            raise ValueError("KmsKeyMetadata requires non-empty 'owner'")

        raw_meta = raw_metadata or {}
        # Enforce strict invariant
        assert_no_private_key_material(raw_meta)

        self.key_id = str(key_id)
        self.algorithm = str(algorithm).upper()
        self.size = size if isinstance(size, int) else str(size)
        self.state = str(state)
        self.owner = str(owner)
        self.provider = str(provider).lower()
        self.description = str(description or "")

        rot = rotation or {}
        self.rotation = {
            "enabled": bool(rot.get("enabled", False)),
            "period_days": int(rot["period_days"]) if rot.get("period_days") is not None else None,
            "last_rotated_at": str(rot["last_rotated_at"]) if rot.get("last_rotated_at") else None,
            "next_rotation_at": str(rot["next_rotation_at"]) if rot.get("next_rotation_at") else None,
            "version": str(rot.get("version", "1")),
        }

        usg = usage or {}
        self.usage = {
            "key_usage": str(usg.get("key_usage", "ENCRYPT_DECRYPT")),
            "operations": list(usg.get("operations", ["encrypt", "decrypt"])),
            "origin": str(usg.get("origin", "KMS")),
            "is_exportable": bool(usg.get("is_exportable", False)),
        }

        self.raw_metadata = sanitize_metadata(raw_meta)

    def to_cbom_component(self) -> Dict[str, Any]:
        """Converts key metadata into CycloneDX CBOM component."""
        try:
            key_len = int(self.size)
        except (ValueError, TypeError):
            key_len = 0

        return {
            "type": "cryptographic-asset",
            "bom-ref": f"cbom:{self.provider}:{self.key_id}",
            "name": self.key_id,
            "description": self.description or f"{self.provider.upper()} Managed Key ({self.algorithm})",
            "cryptoProperties": {
                "assetType": "key",
                "algorithm": self.algorithm,
                "keyLength": key_len,
                "state": self.state,
                "provider": self.provider,
                "rotationEnabled": self.rotation["enabled"],
                "rotationPeriodDays": self.rotation["period_days"],
                "owner": self.owner,
                "keyUsage": self.usage["key_usage"],
                "isExportable": self.usage["is_exportable"],
            },
            "properties": [
                {"name": "ecdat:kms:provider", "value": self.provider},
                {"name": "ecdat:kms:key_id", "value": self.key_id},
                {"name": "ecdat:kms:owner", "value": self.owner},
                {"name": "ecdat:kms:rotation_enabled", "value": str(self.rotation["enabled"]).lower()},
                {"name": "ecdat:kms:state", "value": self.state},
            ],
        }

    def to_dict(self) -> Dict[str, Any]:
        return {
            "key_id": self.key_id,
            "algorithm": self.algorithm,
            "size": self.size,
            "state": self.state,
            "rotation": self.rotation,
            "owner": self.owner,
            "usage": self.usage,
            "provider": self.provider,
            "description": self.description,
            "raw_metadata": self.raw_metadata,
        }
